src/coreclr/jit/fgprofile.cpp

```diff
@@ -2704,29 +2704,19 @@ PhaseStatus Compiler::fgIncorporateProfileData()
 
     if (fgPgoHaveWeights)
     {
-        // We expect not to have both block and edge counts. We may have other
-        // forms of profile data even if we do not have any counts.
+        // If for some reason we have both block and edge counts, prefer the edge counts.
         //
-        // As of 4/6/2023 the following invariant check turns out to no longer hold.
-        // Tracking issue: https://github.com/dotnet/runtime/issues/84446
-        //
-        // assert(!haveBlockCounts || !haveEdgeCounts);
-
         bool dataIsGood = false;
 
-        if (haveBlockCounts)
+        if (haveEdgeCounts)
         {
-            dataIsGood = fgIncorporateBlockCounts();
+            dataIsGood = fgIncorporateEdgeCounts();
         }
-        else if (haveEdgeCounts)
+        else if (haveBlockCounts)
```

Comment:
Would this be safer to also handle the `haveBlockCounts` path if the `dataIsGood` returned from `fgIncorporateEdgeCounts()` was false? That way we would prefer the edge counts, but if they were not incorporable, then we would try using the block counts if we have them...

```csharp
        if (haveEdgeCounts)
        {
            dataIsGood = fgIncorporateEdgeCounts();
        }
        if (!dataIsGood && haveBlockCounts)
        {
            dataIsGood = fgIncorporateBlockCounts();
        }
```

Reply:
That's a reasonable suggestion, but I really don't expect to see block counts anymore, and especially do not expect to see both block and edge counts for the same method -- #80481 enabled edge counts for all cases.

As noted in https://github.com/dotnet/runtime/issues/84446, for some unknown reason the `StandardOptimization.mibc` file that comes from data collected by our internal optimization process (which merges profile data from multiple runs of things) had some cases where a method hand both block and edge counts. We are in the process of updating this profile data (see eg #85193); I should check if that's still the case.